connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java

```diff
@@ -256,39 +277,76 @@ public void testPartitionStateMutation() {
-        assertEquals(newMetadataOffset, partitionState.previousDownstreamOffset, "sync offsets");
-        assertTrue(partitionState.shouldSyncOffsets);
-        verify(producer, times(1)).send(any(), any());
 
-        when(outstandingOffsetSyncs.tryAcquire()).thenReturn(true);
+        assertEquals(recordOffset, partitionState.lastSyncUpstreamOffset, "sync offsets");
+        assertEquals(metadataOffset, partitionState.lastSyncDownstreamOffset, "sync offsets");
+        assertEquals(recordOffset, partitionState.previousUpstreamOffset, "sync offsets");
+        assertEquals(metadataOffset, partitionState.previousDownstreamOffset, "sync offsets");
+        assertFalse(partitionState.shouldSyncOffsets, "partition state reset");
+        verify(producer, times(2)).send(any(), any());
+
+        recordOffset = 4;
+        metadataOffset = 104;
+        recordMetadata = new RecordMetadata(sourceTopicPartition, metadataOffset, 0, 0, 0, recordPartition);
+        sourceRecord = mirrorSourceTask.convertRecord(new ConsumerRecord<>(topicName, recordPartition,
+                recordOffset, System.currentTimeMillis(), TimestampType.CREATE_TIME, recordKey.length,
+                recordValue.length, recordKey, recordValue, headers, Optional.empty()));
+
         mirrorSourceTask.commitRecord(sourceRecord, recordMetadata);
-        assertEquals(newRecordOffset, partitionState.lastSyncUpstreamOffset, "partition state is synced");
-        assertEquals(newMetadataOffset, partitionState.lastSyncDownstreamOffset, "partition state is synced");
-        assertEquals(newRecordOffset, partitionState.previousUpstreamOffset, "partition state is synced");
-        assertEquals(newMetadataOffset, partitionState.previousDownstreamOffset, "partition state is synced");
-        assertFalse(partitionState.shouldSyncOffsets);
+        assertEquals(recordOffset, partitionState.lastSyncUpstreamOffset, "sync offsets");
+        assertEquals(metadataOffset, partitionState.lastSyncDownstreamOffset, "sync offsets");
+        assertEquals(recordOffset, partitionState.previousUpstreamOffset, "sync offsets");
+        assertEquals(metadataOffset, partitionState.previousDownstreamOffset, "sync offsets");
+        assertTrue(partitionState.shouldSyncOffsets, "no partition state reset");
         verify(producer, times(2)).send(any(), any());
+
+        // Should send sync event and should not update last upstream/downstream offset
+        recordOffset = 5;
+        metadataOffset = 150;
+        recordMetadata = new RecordMetadata(sourceTopicPartition, metadataOffset, 0, 0, 0, recordPartition);
+        sourceRecord = mirrorSourceTask.convertRecord(new ConsumerRecord<>(topicName, recordPartition,
+                recordOffset, System.currentTimeMillis(), TimestampType.CREATE_TIME, recordKey.length,
+                recordValue.length, recordKey, recordValue, headers, Optional.empty()));
+
+        outstandingOffsetSyncs.release();
```

Comment:
Using the `ArgumentCaptor` approach:

```suggestion
        producerCallback.getValue().onCompletion(null, null);
```

Reply:
Updated.